### ISSUE TITLE
private method 'inherited' called for Object:Class when using with Rails

### ISSUE THREAD

**Issue by zilkey** [dkubb/axiom#12]
I just created a new rails project like so:

```
rails new veritas-test
```

I then added veritas to the Gemfile, like so:

```
gem 'veritas'
```

I tried to boot that app with `rails c` and got the following error:

```
veritas-test$ rails c
/Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/veritas-0.0.7/lib/veritas/attribute.rb:36:in `inherited': private method `inherited' called for Object:Class (NoMethodError)
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/veritas-0.0.7/lib/veritas/attribute/object.rb:7:in `<class:Attribute>'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/veritas-0.0.7/lib/veritas/attribute/object.rb:4:in `<module:Veritas>'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/veritas-0.0.7/lib/veritas/attribute/object.rb:3:in `<top (required)>'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/veritas-0.0.7/lib/veritas.rb:95:in `require'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/veritas-0.0.7/lib/veritas.rb:95:in `<top (required)>'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/bundler-1.0.22/lib/bundler/runtime.rb:68:in `require'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/bundler-1.0.22/lib/bundler/runtime.rb:68:in `block (2 levels) in require'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/bundler-1.0.22/lib/bundler/runtime.rb:66:in `each'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/bundler-1.0.22/lib/bundler/runtime.rb:66:in `block in require'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/bundler-1.0.22/lib/bundler/runtime.rb:55:in `each'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/bundler-1.0.22/lib/bundler/runtime.rb:55:in `require'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/bundler-1.0.22/lib/bundler.rb:122:in `require'
  from /Users/jeff/workspace/practitionerhq/veritas-test/config/application.rb:7:in `<top (required)>'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/railties-3.2.3/lib/rails/commands.rb:39:in `require'
  from /Users/jeff/.rvm/gems/ruby-1.9.2-p290/gems/railties-3.2.3/lib/rails/commands.rb:39:in `<top (required)>'
  from script/rails:6:in `require'
  from script/rails:6:in `<main>'
```

Any idea what's going on here?  Obviously it's calling inherited on Object, which appears to be private.  If I just load irb, then require rubygems and veritas, it does not blow up like this.

If I can figure this out, and it turns out to be Rails, I'll submit the issue / patch there, but before diving in I wanted to know if this was a known issue or if there was a known workaround.

Thanks!

**Comment by dkubb** [comment]
@zilkey thanks for the bug report, I'll test it out and get back to you.

The odd thing about this is that I've been testing veritas against all versions of ruby since 2009, so it's strange this never came up before. I know that code is being exercised by the tests, because the code has 100% coverage and I'm  using mutation testing -- if those methods weren't being executed the mutations would not cause any specs to fail, which would raise an error.

**Comment by mbj** [comment]
@zilkey I had similar issues with veritas in the past. The presence of active_support with its various core extensions can break veritas (as any other lib). A workarousnd was to ensure veritas is loaded before active_support. I dunno if this strategy will fix this error, but in case you are in hurry you can give it a try.

@dkubb In case my assumption is correct and some of the active_support core extensions causes the problem no code coverage or mutation test can highlight this problem (IMHO). The issue I had was related to active_support patches of `DateTime#<=>` that broke veritas stuff when AS was loaded before.

 BTW I consider AS with the behavior mutation monkey patches as the biggest mistake ever when it comes to reliable developmental under ruby.

**Comment by dkubb** [comment]
@mbj I wonder if there's anything we can do to work around the load order issue? Also I wonder if we should add an integration spec to ensure everything works regardless of which library is loaded first.

**Comment by dkubb** [comment]
Hmm, I see that I fixed this in Virtus (a related gem) a while back: https://github.com/solnic/virtus/issues/14

**Comment by dkubb** [comment]
fwiw, the long term fix is to extract a gem from virtus for doing descendant tracking, and then replace this code with it.

**Comment by mbj** [comment]
@dkubb We can add an integration test that does `Kernel.system("/usr/bin/env/ruby -r active_support -r veritas").should be(true)`.  This would require we have at least a development dependency on active_support. Currently AS is pulled from metric_fu, we should add an explicit development dev in case we add a test like this.

BTW I descendants_tracker is extracted. See https://github.com/dkubb/descendants_tracker. I extracted it from virtus so the fix should already be there.

**Comment by zilkey** [comment]
I just forked the repo, and I was planning to add the dependency on descendents_tracker, but I couldn't get `rake ci` to run - it blows up with `Don't know how to build task 'metrics:heckle'`.  Do you do all development with this on 1.8?  If not, do you have tricks to getting `rake ci` to work on MRI 1.9x?

**Comment by mbj** [comment]
Yeah we do development under 1.8.7 for heckle. You can modify the heckle task to be a noop under ruby-1.9.x. Once [mutant](https://github.com/mbj/mutant) is mature the development will happen under 1.9.

**Comment by dkubb** [comment]
@zilkey I've pushed an update to veritas that uses descendants_tracker which fixes this problem here: https://github.com/dkubb/veritas/commit/0c5f78431a3857fcf0c26f92cf26891787a3b241

BTW the problem you saw with the metrics system is because we run the metrics under 1.8, but we still run the specs under _every_ version of ruby that travis-ci supports. We're kind of forced to keep running the metrics under 1.8 at the moment because the tooling in 1.8 is still superior to what is available under 1.9, rbx or jruby. We're more interested in using the metrics to drive out development and then using 1.9 to verify the behavior works well -- some of us develop in 1.9 and use 1.8 just for verification. Either way, I don't think it matters as long as the ci task passes once in 1.8 prior to pushing the code.

Once @mbj's mutant is ready it will go a long way towards making it so we don't use 1.8 for running the metrics. I'm not yet sure it will completely eliminate the need though, but we'll work on that when we get to it. For now the process we're using is working _really_ well IMHO even with the limitations of having to use 1.8 for final verification of the metrics.

**Comment by mbj** [comment]
@dkubb Yesterday night I was able to fix the problems I had with rspec and the memleaks that occured while preserving the world. We'll see a mutant-0.0.1 today.
